src/pinnable_mapped_file.cpp

```diff
@@ -12,48 +12,99 @@
                                           map_mode mode, std::vector<std::string> hugepage_paths) :
    _data_file_path(bfs::absolute(dir/"shared_memory.bin")),
    _database_name(dir.filename().string()),
    _writable(writable)
 {
-   if(shared_file_size % _db_size_multiple_requirement)
-      BOOST_THROW_EXCEPTION(std::runtime_error("Database must be mulitple of " + std::to_string(_db_size_multiple_requirement) + " bytes"));
+   if(shared_file_size % _db_size_multiple_requirement) {
+      std::cerr << "Database must be mulitple of " <<  std::to_string(_db_size_multiple_requirement) << " bytes" << std::endl;
+      BOOST_THROW_EXCEPTION(std::system_error(make_error_code(db_error_code::bad_size)));
+   }
 #ifndef __linux__
    if(hugepage_paths.size())
-      BOOST_THROW_EXCEPTION(std::runtime_error("Hugepage support is a linux only feature"));
+      BOOST_THROW_EXCEPTION(std::system_error(make_error_code(db_error_code::no_huge_page)));
 #endif
    if(hugepage_paths.size() && mode != locked)
-      BOOST_THROW_EXCEPTION(std::runtime_error("Locked mode is required for hugepage usage"));
+      BOOST_THROW_EXCEPTION(std::system_error(make_error_code(db_error_code::locked_mode_required)));
 #ifdef _WIN32
    if(mode == locked)
-      BOOST_THROW_EXCEPTION(std::runtime_error("Locked mode not supported on win32"));
+      BOOST_THROW_EXCEPTION(std::system_error(make_error_code(db_error_code::no_locked_mode)));
 #endif
 
-   if(!_writable && !bfs::exists(_data_file_path))
-      BOOST_THROW_EXCEPTION(std::runtime_error("database file not found at " + _data_file_path.string()));
+   if(!_writable && !bfs::exists(_data_file_path)){
+      std::cerr << "database file not found at " <<  _data_file_path.string() << std::endl;
+      BOOST_THROW_EXCEPTION(std::system_error(make_error_code(db_error_code::not_found)));
+   }
+
    bfs::create_directories(dir);
 
    if(bfs::exists(_data_file_path)) {
       char header[header_size];
       std::ifstream hs(_data_file_path.generic_string(), std::ifstream::binary);
       hs.read(header, header_size);
       if(hs.fail())
-         BOOST_THROW_EXCEPTION(std::runtime_error("Failed to read DB header."));
+         BOOST_THROW_EXCEPTION(std::system_error(make_error_code(db_error_code::bad_header)));
 
       db_header* dbheader = reinterpret_cast<db_header*>(header);
-      if(dbheader->id != header_id)
-         BOOST_THROW_EXCEPTION(std::runtime_error("\"" + _database_name + "\" database format not compatible with this version of chainbase."));
-      if(!allow_dirty && dbheader->dirty)
-         throw std::runtime_error("\"" + _database_name + "\" database dirty flag set");
+      if(dbheader->id != header_id) {
+         std::cerr << "\"" << _database_name << "\" database format not compatible with this version of chainbase." << std::endl;
+         BOOST_THROW_EXCEPTION(std::system_error(make_error_code(db_error_code::incorrect_db_version)));
```

Comment:
Instead of cerr create a string with this info and pass it into the system_error constructor